Wrappers/Python/cil/optimisation/functions/KullbackLeibler.py

```diff
@@ -1,31 +1,239 @@
+        ind = tmp_sum >= 0
+        tmp = scipy.special.kl_div(self.b.as_array()[ind], tmp_sum[ind])             
+        return numpy.sum(tmp) 
+
+    def gradient(self, x, out = None):
+
+        r"""Returns the value of the gradient of the KullbackLeibler function at :math:`(b, x + \eta)`.                
+        
+        :math:`F'(b, x + \eta) = 1 - \frac{b}{x+\eta}`
+        
+        Note
+        ----
+        To avoid inf values, we :math:`x+\eta>0` is required. 
+        
+        """         
+
+        tmp_sum_array = (x + self.eta).as_array()
+
+        if out is None:   
+            tmp_out = x * 0.0
+            tmp_out.as_array()[tmp_sum_array>0] = \
+                1 - self.b.as_array()[tmp_sum_array>0]/tmp_sum_array[tmp_sum_array>0]
+            return tmp_out
+        else:
+            x.add(self.eta, out=out)
+            out_array = out.clone().as_array()
+            out_array[tmp_sum_array>0] = 1 - self.b.as_array()[tmp_sum_array>0]/tmp_sum_array[tmp_sum_array>0]
+            out.fill(out_array)
+
+
+    def convex_conjugate(self, x):
+
+        r"""Returns the value of the convex conjugate of the KullbackLeibler function at :math:`(b, x + \eta)`.                
+        
+        :math:`F^{*}(b, x + \eta) = - b \log(1-x^{*}) - <x^{*}, \eta>`
+        
+        """          
 
+        tmp = 1 - x.as_array()
+        ind = tmp>0
+        xlogy = - scipy.special.xlogy(self.b.as_array()[ind], tmp[ind])
+        return numpy.sum(xlogy) - self.eta.dot(x)
+
+    def proximal(self, x, tau, out = None):
+
+        r"""Returns the value of the proximal operator of the KullbackLeibler function at :math:`(b, x + \eta)`.
+        
+        :math:`\mathrm{prox}_{\tau F}(x) = \frac{1}{2}\bigg( (x - \eta - \tau) + \sqrt{ (x + \eta - \tau)^2 + 4\tau b} \bigg)`
+                            
+        """        
+        
+        if out is None:        
+            return 0.5 *( (x - self.eta - tau) + \
+                ( (x + self.eta - tau).power(2) + 4*tau*self.b   ) .sqrt() \
+                    )        
+        else:                      
+            x.add(self.eta, out=out)
+            out -= tau
+            out *= out
+            out.add(self.b * (4 * tau), out=out)
+            out.sqrt(out=out)  
+            out.subtract(tau, out=out)
+            out.subtract(self.eta, out=out)
+            out.add(x, out=out)         
+            out *= 0.5             
+
+
+    def proximal_conjugate(self, x, tau, out = None):
+
+        r"""Returns the value of the proximal operator of the convex conjugate of KullbackLeibler at :math:`(b, x + \eta)`.
+           
+        :math:`\mathrm{prox}_{\tau F^{*}}(x) = 0.5*((z + 1) - \sqrt{(z-1)^2 + 4 * \tau b})`, where :math:`z = x + \tau \eta`.
+        """        
+        
+        if out is None:
+            z = x + tau * self.eta
+            return 0.5*((z + 1) - ((z-1).power(2) + 4 * tau * self.b).sqrt())
+        else:            
+            tmp = tau * self.eta
```

Comment:
use `sapyb`

Reply:
Postponed for the next version, see #1317